sql/database.go

```diff
@@ -511,6 +538,20 @@ func (tx *Tx) Exec(query string, encoder Encoder, decoder Decoder) (int, error)
 	return exec(tx.conn, query, encoder, decoder)
 }
 
+func mapSqliteError(err error) error {
+	code := sqlite.ErrCode(err)
+	if code == sqlite.SQLITE_CONSTRAINT_PRIMARYKEY || code == sqlite.SQLITE_CONSTRAINT_UNIQUE {
+		return ErrObjectExists
+	}
+	if code == sqlite.SQLITE_INTERRUPT {
+		// TODO: we probably should check if there was indeed a context
+		// that was canceled. But we're likely to replace crawshaw library
+		// in future so this part should be rewritten anyway
+		return context.Canceled
+	}
+	return err
+}
```

Comment:
I don't think returning a `context.Canceled` with the assumption that if `SQLITE_INTERRUPT` happened a context must have been canceled somewhere is a good idea. From what I can see the returned error usually doesn't happen in a codepath that even is context aware.

Also I think its easier to parse what's happening here with a `switch`/`case` statement:

```suggestion
func mapSqliteError(err error) error {
	switch sqlite.ErrCode(err) {
	case sqlite.SQLITE_CONSTRAINT_PRIMARYKEY:
		return ErrObjectExists
	case sqlite.SQLITE_CONSTRAINT_UNIQUE:
		return ErrObjectExists
	default:
		return err
	}
}
```

Reply:
The code path that leads to this is the following:
1. When a transaction is initialized, a `context.Context` is passed to e.g. `Database.WithTx`
2. The context is used to acquire an SQLite connection from the pool
3. During this, the connection is set to be interrupted when the context is canceled: https://github.com/go-llsqlite/crawshaw/blob/v0.5.3/sqlitex/pool.go#L177
4. Without this extra check, this interruption during a long-running SQL query leads to a confusing error message that obscures the fact that the cause was context cancellation

I actually did see this error on real nodes and thus I added extra check for this interruption.
The proper way to deal with it would be checking whether the actual context used to start the transaction is canceled in addition to checking the error code, but that would involve some extra modifications which I'd prefer to do in a separate PR.
So, I'm converting this code to a `switch`, but keeping the `sqlite.SQLITE_INTERRUPT`; also removing the comment about the need to replace `crawshaw`, as we decided against it in a separate discussion

Reply:
Issue created: #6273
There are further concerns mentioned in the issue about the fact that interrupting any writes to the DB will not rollback the transaction and will cause deadlocks later on.

Reply:
Regarding the issue with a transaction not being rolled back on ctx cancellation: go 1.21 introduced [context.AfterFunc](https://pkg.go.dev/context#AfterFunc). Witin `WithTx` we could add `tx.Release` as an `AfterFunc` that gets executed when the context is cancelled and releases/rolls back the transaction. Wdyt?

Reply:
This will likely help, but it doesn't fully resolve the related write deadlock issue, and also there's still need to interrupt a long-running query which will always result in `sqlite.SQLITE_INTERRUPT`. In any case, I think it makes sense to do such changes in a another PR, as this one is quite large and long-lasting already, on top of being in the need of constant resolution of merge conflicts